### ISSUE TITLE
func-names incorrectly errors on get/set/shorthand methods.

### ISSUE THREAD

**Issue by keithamus** [eslint/eslint#1699]
### Abstract

`func-names` rule errors for valid getter/setter and es6 shorthand functions

### Details

Given the following JS:

```js
exports = {
  get foo() {
     return 1;
   }
 };
```

This produces the following Esprima syntax tree (trimmed for brevity & relevancy):

```json
 "properties": [{
    "type": "Property",
    "key": {
        "type": "Identifier",
        "name": "foo"
    },
    "value": {
        "type": "FunctionExpression",
        "id": null,
        "params": [],
        "defaults": [],
        "body": {
            "type": "BlockStatement",
            "body": []
        },
        "rest": null,
        "generator": false,
        "expression": false
    },
    "kind": "get",
    "method": false,
    "shorthand": false
}]
```

Given the following ES6 JS:

```js
exports = {
    foo() {
        return 1;
    }
};
```

This produces the following Esprima syntax tree (trimmed for brevity & relevancy):

```json
"properties": [{
    "type": "Property",
    "key": {
        "type": "Identifier",
        "name": "foo"
    },
    "value": {
        "type": "FunctionExpression",
        "id": null,
        "params": [],
        "defaults": [],
        "body": {
            "type": "BlockStatement",
            "body": [{
                "type": "ReturnStatement",
                "argument": {
                    "type": "Literal",
                    "value": 1,
                    "raw": "1"
                }
            }]
        },
        "rest": null,
        "generator": false,
        "expression": false
    },
    "kind": "init",
    "method": true,
    "shorthand": false
}]
```

func-names looks for `FunctionExpression` tokens, and attempts to ensure the expression has a name. However, in the case of a getter/setter/shorthand it doesn't (the name is the property name). This means func-names errors on this code snippet with `error  Missing function expression name`.

### Solution

`func-names` needs to check the `kind` flag. If it is `'get'`, `'set'` or `'init'` then func-names should not attempt to validate the name.

**Comment by nzakas** [comment]
Quite possibly the best bug report ever. Thanks, we will fix this.

**Comment by btmills** [comment]
This will also apply to object literal shorthand methods:

```js
exports = {
    foo() {
        return 1;
    }
};
```

While `Property.kind` will still be `"init"`, `Property.method` will be `true` in this case.

Working on this.

**Comment by keithamus** [comment]
Totally right @btmills. OP updated to reflect your additions.
